src/lib/components/Modal/_hooks/useModalFocus.js

```diff
@@ -0,0 +1,96 @@
+import { useEffect } from 'react';
+
+export const useModalFocus = (
+  autoFocus,
+  childrenWrapperRef,
+  closeButtonRef,
+  focusTrap,
+  primaryButtonRef,
+) => {
+  useEffect(
+    () => {
+      if (!autoFocus) {
+        return () => {};
+      }
+
+      // If `autoFocus` is set to `true`, following code finds all focusable elements and among them
+      // first not disabled form field element (input, textarea or select) or primary button
+      // and auto focuses it. This is necessary to have focus on one of those elements to be able
+      // to submit the form by pressing Enter key. If there are neither, it tries to focus any other
+      // focusable elements. In case there are none, childrenWrapperElement (Modal itself) is focused.
+
+      const childrenWrapperElement = childrenWrapperRef.current;
+
+      if (childrenWrapperElement == null) {
+        return () => {};
+      }
+
+      const childrenFocusableElements = Array.from(
+        childrenWrapperElement.querySelectorAll('button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'),
```

Comment:
Are we sure that `'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'` is final list of focusable elements? It is more question that my requirement, but what about allowing user to override this list of elements?

Something like this:

```
Modal.defaultProps = {
  autoFocusableElemenes: [
    'button',
    '[href]',
    'input',
    ...
  ]
}
```

cc @mbohal @adamkudrna @hubacekj

Reply:
I would leave it as is.

If we come across something generally useful that we forgot we can add it to the list.

If we come across a use case when we will need to focus something special that is not worth adding to the list I would prefer passing to the `Modal` in a React `Ref` pointing to the element to be focused on `Modal` open (something like `initialFocusRef`). That is in my view more flexible.

I would not worry about it now though.